### PR TITLE
🏗✨ Upgrade closure compiler to v20190513

### PR DESCRIPTION
**Pull request by rsimha**
This PR upgrades the version of closure compiler used to minify the AMP runtime to [`v20190513`](https://github.com/google/closure-compiler/wiki/Releases#may-13-2019-v20190513).

**Highlights:**
- Updated `compiler.jar` with a custom built version based on `v20190513`
    - Reverts two breaking changes in closure compiler code ([one](https://github.com/google/closure-compiler/commit/0393c80ca01b6b861376dad7f91043a38bb37dc0), [two](https://github.com/google/closure-compiler/commit/acea045bcc566c8d1ba74bc8f63fbb1568a18c2b)) (required until https://github.com/google/closure-compiler/issues/3041 is fixed)
- Updated `compiler-and-tests.jar` with a custom built version based on `v20190513`
    - Follows the steps outlined [here](https://github.com/ampproject/amphtml/pull/18552#issue-220501644)
- Updated documentation for custom build changes in [README.amp](https://github.com/rsimha/amphtml/blob/2019-05-21-ClosureCompiler/third_party/closure-compiler/README.amp)
- Updated `runner.jar` after fixing code and tests to be compatible with latest compiler code
- Fixed a few type check errors discovered by the new version
- Adopted the [recently introduced](https://github.com/google/closure-library/commit/192e56847eb8673475e21cff58db364b392345e1) `@closurePrimitive` [annotations](https://github.com/google/closure-compiler/blob/maven-v20190513/src/com/google/javascript/rhino/ClosurePrimitive.java#L45-L59) for assertion functions
- Removed the `getAssertionFunctions()` [override](https://github.com/ampproject/amphtml/blob/319257b64603283f1e642250f4dfff14012e98ea/build-system/runner/src/org/ampproject/AmpCodingConvention.java#L48-L66) in `AmpCodingConvention.java`

**Future work:** Dynamically generate `runner.jar` once we no longer need a custom built `compiler.jar` (#22452)

Companion to #22368
Partial fix for #18748
Follow up to #18552, #18609, #18794, #19449, #20056, and #21618

### CODE REVIEW

File: build-system/runner/src/org/ampproject/AmpCodingConvention.java [hunk 1]
```diff
@@ -48,20 +47,20 @@ public AmpCodingConvention(CodingConvention convention) {
...
-          "module$src$log.userAssert"),
-      AssertionFunctionSpec.makeReturnTypeAssertion(
-          "userAssert$$module$src$log"),
-      AssertionFunctionSpec.makeReturnTypeAssertion(
-          "assertService$$module$src$element_service"),
-      AssertionFunctionSpec.makeReturnTypeAssertion(
-          "module$src$layout.assertLength"),
-      AssertionFunctionSpec.makeReturnTypeAssertion(
-          "assertLength$$module$src$layout")
+      AssertionFunctionSpec.forMatchesReturn().setFunctionName(
```
Comment (jridgewell) [inline build-system/runner/src/org/ampproject/AmpCodingConvention.java]: If we're upgrading Closure Compiler, can we switch to https://github.com/google/closure-library/blob/master/closure/goog/asserts/asserts.js#L181?

Comment (rsimha) [reply build-system/runner/src/org/ampproject/AmpCodingConvention.java]: Good idea. Trying this. Shouldn't these be `@closurePrimitive {asserts.matchesReturn}` though?

https://github.com/google/closure-library/blob/master/closure/goog/asserts/asserts.js#L212

Comment (jridgewell) [reply build-system/runner/src/org/ampproject/AmpCodingConvention.java]: It appears so https://github.com/google/closure-compiler/blob/004024fce9c307e8855c8eaa376d556fb344e763/src/com/google/javascript/rhino/ClosurePrimitive.java#L57-L59

Comment (rsimha) [reply build-system/runner/src/org/ampproject/AmpCodingConvention.java]: Yep, I see the same set of options in the [v20190513](https://github.com/google/closure-compiler/blob/maven-v20190513/src/com/google/javascript/rhino/ClosurePrimitive.java#L57-L59) codebase. Trying this.

Comment (rsimha) [reply build-system/runner/src/org/ampproject/AmpCodingConvention.java]: That worked! `v0.js` is exactly the same before and after this change. See 5ebcee3cd569d11295cc8f0dda1ee0a17c7b222c.

**Comment by rsimha** [comment]
**Tested this as follows:**
- Running time for `gulp dist` is unaffected ([link](https://travis-ci.org/ampproject/amphtml/jobs/536008630#L386))
- Zero visual diffs ([link](https://travis-ci.org/ampproject/amphtml/jobs/536008637#L2014))
- Single pass build and tests are green ([link](https://travis-ci.org/ampproject/amphtml/jobs/536008635#L396))
- Local tests and remote tests all green ([link](https://travis-ci.org/ampproject/amphtml/jobs/536008639), [link](https://travis-ci.org/ampproject/amphtml/jobs/536008641))
- Number of instances of the string `assert` in `v0.js` remains unchanged in spite of 2a45d76c4cc3a7da55fa72bcabc294a80404a2fa
- Manually sanity checked the runtime + extensions in multi pass and single pass mode
- Source maps continue to work with this [verification tool](http://sokra.github.io/source-map-visualization/#)

**Pull request merged**